### DISCUSSION TITLE
Multiple outputs

### DISCUSSION THREAD

**Discussion by c0nsoleg** [joe-bell/cva#94] [Ideas]
Hi,

I have encountered many places where component, for example some banner, need adjustments on more than one structure level but is based on same params. Right now we can create multiple cva and keep needed variants consistent but what if we could support multiple output within single cva and single variant names.

For example, instead of creating two separate cva with same variants:
```
const containerStyles = cva([], {
  variants: {
    size: {
      wide: [],
      base: []
    }
  }
})
const buttonStyles = cva([], {
  variants: {
    size: {
      wide: [],
      base: []
    }
  }
})
```

we could write it like that
```
const styles = cva({container: [], button: [] }, {
  variants: {
    wide: {
      container: [],
      button: []
    }
  }
})
```

Many thanks,
Mateusz

**Comment by joe-bell** [comment]
~I think it's fairly unlikely I'll add support for this in the near-future, but you can achieve this by "composing components" in the meantime:~

~https://github.com/joe-bell/cva#composing-components~

Misunderstood the question, apologies!

https://github.com/joe-bell/cva/discussions/94#discussioncomment-5009225

**Reply by bwong365** [reply]
> you can achieve this by "composing components" in the meantime:

Hey @joe-bell if I'm not mistaken, the docs discuss extension being possible if you want to style a single component with concatenated cva functions, as opposed to being able to output multiple classes to target the different parts of a compound component like a headless ui Listbox (https://headlessui.com/react/listbox). Is there a recommended strategy for that sort of thing right now?

**Reply by joe-bell** [reply]
Aha gotcha! I like this idea. I'll investigate for `cva@1.0.0`

**Reply by joe-bell** [reply]
Just dropping back in to say this will be considered for `cva@1.0`

Thank you for your suggestion 🖤

**Reply by NickWoodward** [reply]
Definitely keen on this - is there a workaround until then? Like I've a Navigation Item that has an icon in it, and I'd like not only the styles of the NavigationItem to be conditional on the props, but the icon's wrapper too:

```

const itemVariants = cva("flex w-full border-b border-gray-400", {
  variants: {
    variant: {
      default: //...,
      mobile: //...,
    },
    size: {
      md: //...,
      lg: //...
    },
  },
  defaultVariants: {
    variant: "mobile",
    size: "md",
  },
});

export const NavigationItem = ({
  linkClasses,
  item,
  variant,
  size,
}: NavItemProps) => {
  return (
    <NavLink
      key={item.id}
      to={item.to}
      className={cn(itemVariants({ variant, size }), linkClasses)}
    >
      <div>{item.icon}</div>
      {item.label}
    </NavLink>
  );
};

```

Thanks!

**Reply by joe-bell** [reply]
I would recommend leveraging arbitrary TailwindCSS styles to set some CSS variables – CSS already solves this problem pretty nicely!

**Reply by NickWoodward** [reply]
Hi Joe! Appreciate the quick reply and love the package.

I'm not quite sure what you mean though. Definitely my fault, but wouldn't I still have to conditionally apply or change those variables?

For example if I want a `size="sm"` NavigationItem to always have a small icon, or a `size="lg"` NavigationItem to have a medium icon, how could I achieve that with css variables?

**Reply by c0nsoleg** [reply]
Maybe by using something like `[&_svg]:text-sm`? or any other class after :

**Reply by joe-bell** [reply]
This PR is a work in progress – docs coming shortly – but hope this helps 🙏 

https://github.com/joe-bell/cva/blob/feat/compound/examples/react-with-tailwindcss-compound/src/components/nav/nav.tsx

**Reply by NickWoodward** [reply]
Hi c0nsoleg

Not quite sure how I can use that to make the icon change based on the parent's prop though?

At the moment I'm thinking of flattening the NavigationItem and creating a new Icon component that also uses CVA, but I'm also a little worried about bloat

```
<NavigationItem size="sm" icon={<Icon size="lg" />} />
```

so this way I can just manually decide the Icon size when I use the NavigationItem

**Reply by NickWoodward** [reply]
> This PR is a work in progress – docs coming shortly – but hope this helps 🙏
> 
> https://github.com/joe-bell/cva/blob/feat/compound/examples/react-with-tailwindcss-compound/src/components/nav/nav.tsx

Thanks a lot!

**Reply by ngekoding** [reply]
Sorry, can we use it also in Vue?

**Comment by mnieber** [comment]
I've created an approach to solve this issue, not knowing that cva existed (so I've partly reinvented the wheel). 

I believe that we should not use `className` to pass in the styles, but rather pass in a `trim` object. Then from inside the component, different parts of the trim object are used to style different parts of the component.

```
export type DialogButtonTrimT = {
  base: {
    root: {
      border: any;
      color: any;
      fontSize: any;
      padding: any;
    };
    Icon: {
      margin: any;
      size: any;
      color: any;
    };
  };
  danger?: ModeOverlayT<DialogButtonTrimT>;
  disabled?: ModeOverlayT<DialogButtonTrimT>;
};

export type PropsT = {
  className?: any;
  trim: DialogButtonTrimT;
  danger?: boolean;
  disabled?: boolean;
};

export const DialogButton = (props: PropsT) => {
  // style is determined by the trim prop:
  // - root style is based on trim.base.root
  // - the button has an icon that uses a style based on trim.base.Icon.
  // if the button is in disabled or danger mode, then the disabled or danger overlay is used to augment the base style.
}
```

The contents of the trim object (i.e. the actual tailwindcss styles) come from a style-object. These are all minor differences from cva (the approach is basically the same), but using `trim` instead of `className` offers some advantages I think.

If you want to discuss, please let me know.

ps Link to my blog post regarding this topic: https://maartennieber.com/posts/styling-system-style-objects-scss/chapter-4
